File: src/main/java/name/gyger/jmoney/category/CategoryService.java
package name.gyger.jmoney.category;

import name.gyger.jmoney.session.Session;
import name.gyger.jmoney.session.SessionService;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.Query;
import java.util.ArrayList;
import java.util.List;

@Service
@Transactional
public class CategoryService {

    private final SessionService sessionService;

    @PersistenceContext
    private EntityManager em;

    public CategoryService(SessionService sessionService) {
        this.sessionService = sessionService;
    }

    public List<Category> prefetchCategories() {
        Query q = em.createQuery("SELECT c FROM Category c LEFT JOIN FETCH c.children", Category.class);
        return (List<Category>) q.getResultList();
    }

    public Category getRootCategory() {
        Session s = sessionService.getSession();
        prefetchCategories();
        return s.getRootCategory();
    }

    public List<Category> getCategories() {
        Category rootCategory = getRootCategory();
        List<Category> categories = new ArrayList<>();
        if (rootCategory != null) {
            addChildCategories(categories, rootCategory, 0);
        }
        categories.forEach(c -> {
            em.detach(c);
            Category parent = c.getParent();
            if (parent != null) c.setParentId(parent.getId());
        });
        return categories;
    }

    private void addChildCategories(List<Category> categories, Category parentCategory, int level) {
        for (Category childCategory : parentCategory.getChildren()) {
            childCategory.setLevel(level);
            categories.add(childCategory);
            addChildCategories(categories, childCategory, level + 1);
        }
    }

    public Category getCategoryTree() {
        Category root = getRootCategory();
        resolveChildCats(root, 0);
        return root;
    }

    private void resolveChildCats(Category node, int level) {
        Category parent = node.getParent();
        node.setLevel(level);
        if (parent != null) node.setParentId(parent.getId());
        node.getChildren().forEach(c -> resolveChildCats(c, level + 1));
    }

    public void saveCategoryTree(Category category) {
        resolveParents(category);
        em.merge(category);
    }

    private void resolveParents(Category node) {
        Category parent = em.find(Category.class, node.getParentId());
        node.setParent(parent);
        node.getChildren().forEach(c -> resolveParents(c));
    }

    public long createCategory(Category category) {
        Category parent = em.find(Category.class, category.getParentId());
        category.setParent(parent);
        em.persist(category);
        return category.getId();
    }

    public void deleteCategory(long categoryId) {
        Query q = em.createNativeQuery("UPDATE ENTRY SET CATEGORY_ID = NULL WHERE CATEGORY_ID = :categoryId");
        q.setParameter("categoryId", categoryId);
        q.executeUpdate();

        Category category = em.find(Category.class, categoryId);
        em.remove(category);
    }

    public Category getSplitCategory() {
        return sessionService.getSession().getSplitCategory();
    }

    public Category getTransferCategory() {
        return sessionService.getSession().getTransferCategory();
    }

}

File: src/main/kotlin/name/gyger/jmoney/category/CategoryService.kt
package name.gyger.jmoney.category

import name.gyger.jmoney.account.EntryRepository
import name.gyger.jmoney.session.SessionRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
open class CategoryService(private val sessionRepository: SessionRepository,
                           private val categoryRepository: CategoryRepository,
                           private val entryRepository: EntryRepository) {

    fun getRootCategory(): Category {
        categoryRepository.findAll()  // prefetch
        return sessionRepository.getSession().rootCategory
    }

    fun getCategories(): List<Category> {
        val categories = ArrayList<Category>()
        addChildCategories(categories, getRootCategory(), 0)
        categories.forEach { c ->
            c.parentId = c.parent?.id ?: 0
        }
        return categories
    }

    private fun addChildCategories(categories: MutableList<Category>, parentCategory: Category, level: Int) {
        parentCategory.children.forEach {
            it.level = level
            categories.add(it)
            addChildCategories(categories, it, level + 1)
        }
    }

    fun getCategoryTree(): Category {
        val root = getRootCategory()
        resolveChildCats(root, 0)
        return root
    }

    private fun resolveChildCats(node: Category, level: Int) {
        val parent = node.parent
        node.level = level
        if (parent != null) node.parentId = parent.id
        node.children.forEach { c -> resolveChildCats(c, level + 1) }
    }

    fun saveCategoryTree(category: Category) {
        resolveParents(category)
        categoryRepository.save(category)
    }

    private fun resolveParents(node: Category) {
        node.parent = categoryRepository.findOne(node.parentId)
        node.children.forEach { c -> resolveParents(c) }
    }

    fun createCategory(category: Category): Long {
        category.parent = categoryRepository.findOne(category.parentId)
        categoryRepository.save(category)
        return category.id
    }

    fun deleteCategory(categoryId: Long) {
        entryRepository.deleteCategoryFromEntry(categoryId)
        categoryRepository.delete(categoryId)
    }

    fun getSplitCategory(): Category {
        return sessionRepository.getSession().splitCategory
    }

    fun getTransferCategory(): Category {
        return sessionRepository.getSession().transferCategory
    }

}
